File: go/unisig.go
package unisig

import (
	"errors"
	"os"
)

const magic = "\xDC\xDC\x0D\x0A\x1A\x0A\x00"

func Read(f os.File) (string, bool, error) {
	head := make([]byte, len(magic)+1)
	n, err := f.Read(head)
	if err != nil {
		return "", false, err
	}
	if n != len(head) {
		return "", false, errors.New("short read")
	}
	lenbyte := head[len(magic)]
	isUUID := (lenbyte == 0)
	length := int(lenbyte)
	if isUUID {
		length = 16
	}
	sig := make([]byte, length)
	n, err = f.Read(sig)
	if err != nil {
		return "", false, err
	}
	if n != length {
		return "", false, errors.New("short read")
	}
	return string(sig), isUUID, nil
}

func WriteURI(f os.File, uri string) error {
	if len(uri) > 255 {
		return errors.New("URI too long")
	}
	whole := append([]byte(magic), byte(len(uri)))
	whole = append(whole, uri...)
	_, err := f.Write(whole)
	return err
}

func WriteUUID(f os.File, uuid string) error {
	if len(uuid) != 16 {
		return errors.New("UUID not 16 bytes")
	}
	whole := append([]byte(magic), 0)
	whole = append(whole, []byte(uuid)...)
	_, err := f.Write(whole)
	return err
}

File: ruby/unisig.rb
MAGIC = [0xDC, 0xDC, 0x0D, 0x0A, 0x1A, 0x0A, 0x00].pack("C*")

def read(io)
  raise unless io.binmode?
  head = io.read(MAGIC.length + 1)
  if head.length != MAGIC.length + 1
    return false
  end
  if head.slice(0, MAGIC.length) != MAGIC
    return false
  end
  lenbyte = head[MAGIC.length].ord
  is_uuid = (lenbyte == 0)
  len = (if is_uuid then 16 else lenbyte end)
  sig = io.read(len)
  return false if sig.length != len
  if is_uuid then sig.unpack("C*") else sig.force_encoding("UTF-8") end
end

def write(io, name)
  raise unless name.length <= 255
  raise unless io.binmode?
  if name.is_a?(Array)
    raise unless name.length == 16
    sig = name
    is_uuid = true
  else
    raise unless name.is_a?(String)
    sig = name.unpack("utf-8", "U*")
    is_uuid = false
  end
  io.write(MAGIC)
  io.write(bytes([if is_uuid then 0 else sig.length end]))
  io.write(sig)
end
